hands-on-exercise/1-native/2-build-module.md

```diff
@@ -0,0 +1,980 @@
+#!/usr/bin/make -f
+
+DOCKER := $(shell which docker)
+
+#################
+###   Build   ###
+#################
+
+test:
+    @echo "--> Running tests"
+    go test -v ./...
+
+test-integration:
+    @echo "--> Running integration tests"
+    cd integration; go test -v ./...
+
+.PHONY: test test-integration
+
+##################
+###  Protobuf  ###
+##################
+
+protoVer=0.14.0
+protoImageName=ghcr.io/cosmos/proto-builder:$(protoVer)
+protoImage=$(DOCKER) run --rm -v $(CURDIR):/workspace --workdir /workspace $(protoImageName)
+
+proto-all: proto-format proto-lint proto-gen
+
+proto-gen:
+    @echo "Generating protobuf files..."
+    @$(protoImage) sh ./scripts/protocgen.sh
+    @go mod tidy
+
+proto-format:
+    @$(protoImage) find ./ -name "*.proto" -exec clang-format -i {} \;
+
+proto-lint:
+    @$(protoImage) buf lint proto/ --error-format=json
+
+.PHONY: proto-all proto-gen proto-format proto-lint
+
+#################
+###  Linting  ###
+#################
+
+golangci_lint_cmd=golangci-lint
+golangci_version=v1.51.2
+
+lint:
+    @echo "--> Running linter"
+    @go install github.com/golangci/golangci-lint/cmd/golangci-lint@$(golangci_version)
+    @$(golangci_lint_cmd) run ./... --timeout 15m
+
+lint-fix:
+    @echo "--> Running linter and fixing issues"
+    @go install github.com/golangci/golangci-lint/cmd/golangci-lint@$(golangci_version)
+    @$(golangci_lint_cmd) run ./... --fix --timeout 15m
+
+.PHONY: lint lint-fix
+```
+
+</ExpansionPanel>
+
+Note how it uses the [`ghcr.io/cosmos/proto-builder:0.14.0`](https://github.com/cosmos/cosmos-sdk/pkgs/container/proto-builder/119928846?tag=0.14.0) Docker image here to run the `protocgen.sh` script. This helps with making sure all necessary software is available.
+
+### First Protobuf compilation
+
+Before you run the compilation, because you have not yet defined any Protobuf files for messages, you can comment out this `protocgen.sh` line:
+
+```diff-sh
+-   cp -r github.com/alice/checkers/* ./
++   # cp -r github.com/alice/checkers/* ./
+```
+
+Now run the compilation:
+
+```sh
+$ make proto-gen
+```
+
+This creates a new `api/module/v1/module.pulsar.go` file and updates your `go.mod`. It also creates a new `go.sum`.
+
+Now you need a couple more Protobuf files.
+
+### Minimum Protobuf objects
+
+You are not defining your games, messages, and queries just yet. However, you already define the facts that your module:
+
+* Has a genesis, which has a type.
+* Uses params, which also have a type.
+
+Go ahead and define them in a new `proto/alice/checkers/v1/types.proto`, taking inspiration from `minimal-module-example`:
+
+```protobuf [proto/alice/checkers/v1/types.proto]
+syntax = "proto3";
+package alice.checkers.v1;
+
+option go_package = "github.com/alice/checkers";
+
+import "cosmos_proto/cosmos.proto";
+import "gogoproto/gogo.proto";
+import "amino/amino.proto";
+
+// Params defines the parameters of the module.
+message Params { option (amino.name) = "alice/checkers/Params"; }
```

Comment:
amino name is not required for new modules, it is meant to help with byte for byte compatibility with existing modules.

Reply:
I see that Julien [did that part](https://github.com/cosmosregistry/example/commit/16fd24ff75a6f76bb5a8f8a08df9e099b0d2f10b) already, so now I have my example.